### ISSUE TITLE
compression+encryption leads to multiple bugs

### ISSUE THREAD

**Issue by harshavardhana** [minio/minio#11447]
## Expected Behavior
All mint tests run fine exactly as with encrypted deployment

## Current Behavior
mint fails for multiple reasons 

### Returned md5sum mismatches, we need to preserve md5sum or add `-1` at the end if it's not expected to be same
```
(5/15) Running awscli tests ... FAILED in 32 seconds
{
  "name": "awscli",
  "duration": 2436,
  "function": "aws --endpoint-url http://minio-dev7.minio.io:32627 s3api copy-object --bucket awscli-mint-test-bucket-29930 --key datafile-1-kB-copy --copy-source awscli-mint-test-bucket-29930/datafile-1-kB\n",
  "status": "FAIL",
  "error": "Hash mismatch expected 084e1383b70fb0c51acc680fef370023, got ac57de7156d7fc25ac1a65f81fa3989b"
}
```

### minio-go functional tests UnencryptedToSSE3CopyObject fails

```
(8/15) Running minio-go tests ... FAILED in 2 minutes and 7 seconds
{
  "args": {
    "destination": {
      "Bucket": "minio-go-test-ul2sejgj5p3aojuz",
      "Object": "dstObject",
      "Encryption": {},
      "UserMetadata": null,
      "ReplaceMetadata": false,
      "UserTags": null,
      "ReplaceTags": false,
      "LegalHold": "",
      "Mode": "",
      "RetainUntilDate": "0001-01-01T00:00:00Z",
      "Size": 0,
      "Progress": null
    },
    "source": {
      "Bucket": "minio-go-test-ul2sejgj5p3aojuz",
      "Object": "srcObject",
      "VersionID": "",
      "MatchETag": "",
      "NoMatchETag": "",
      "MatchModifiedSince": "0001-01-01T00:00:00Z",
      "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
      "MatchRange": false,
      "Start": 0,
      "End": 0,
      "Encryption": null
    }
  },
  "duration": 4878,
  "error": "We encountered an internal error, please try again.: cause(s2: corrupt input)",
  "function": "CopyObject(destination, source)",
  "message": "GetObject failed",
  "name": "minio-go: testUnencryptedToSSES3CopyObject",
  "status": "FAIL"
}
```

### Preconditions for matching etag fails with minio-java

```

(8/15) Running minio-java tests ... FAILED in 1 minutes and 30 seconds
{
  "name": "minio-java",
  "function": "copyObject()",
  "args": "[match etag]",
  "duration": 316,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = PreconditionFailed, message = At least one of the pre-conditions you specified did not hold, bucketName = minio-java-test-qm42nc, objectName 
= minio-java-test-34746a5-copy, resource = /minio-java-test-qm42nc/minio-java-test-34746a5-copy, requestId = 1660A3D80101FE14, hostId = 6c6b4eef-e46f-462b-9379-8152149650dd)\nrequest={method
=PUT, url=http://minio-dev7.minio.io:32627/minio-java-test-qm42nc/minio-java-test-34746a5-copy, headers=x-amz-copy-source-if-match: 71cff0a060f852067e443ad1e24ae26c-1\nx-amz-copy-source: /mi
nio-java-test-1uvg7v2/minio-java-test-34746a5\nHost: minio-dev7.minio.io:32627\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.0.3\nContent-MD5: 1B2M2Y8AsgTpgAmY7Ph
Cfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20210204T200014Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20210204/us
-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-copy-source;x-amz-copy-source-if-match;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=412, headers=Acc
ept-Ranges: bytes\nContent-Length: 416\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nETag: \"71cff0a060f852067e443ad1e24ae26c\"\nLast-Modified: Thu, 04 Fe
b 2021 20:00:14 GMT\nServer: MinIO\nVary: Origin\nX-Amz-Request-Id: 1660A3D80101FE14\nX-Xss-Protection: 1; mode=block\nDate: Thu, 04 Feb 2021 20:00:14 GMT\n}\n >>> [io.minio.MinioClient.exec
ute(MinioClient.java:775), io.minio.MinioClient.execute(MinioClient.java:563), io.minio.MinioClient.executePut(MinioClient.java:904), io.minio.MinioClient.copyObject(MinioClient.java:1232), 
FunctionalTest.testCopyObjectMatchETag(FunctionalTest.java:1850), FunctionalTest.copyObject(FunctionalTest.java:2016), FunctionalTest.runObjectTests(FunctionalTest.java:3757), FunctionalTest
.runTests(FunctionalTest.java:3783), FunctionalTest.main(FunctionalTest.java:3927)]"
}
```

### MD5sum does not match with s3cmd
```
(10/15) Running s3cmd tests ... FAILED in 6 seconds
{
  "name": "s3cmd",
  "duration": "2914",
  "function": "test_put_object_multipart",
  "status": "FAIL",
  "error": "WARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARN
ING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums do
n't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Too many failures. Giving up on '/mint/data/datafile-65-MB'\nERROR: \nUpload of '/
mint/data/datafile-65-MB' part 1 failed. Use\n  /usr/local/bin/s3cmd abortmp s3://s3cmd-test-bucket-21467/s3cmd-test-object-20748 72223042-a1cd-4094-a6f1-b9c4af811448\nto abort the upload, o
r\n  /usr/local/bin/s3cmd --upload-id 72223042-a1cd-4094-a6f1-b9c4af811448 put ...\nto continue the upload.\nERROR: Upload of '/mint/data/datafile-65-MB' failed too many times (Last reason: 
)"
}
```
## Possible Solution
- Just like how we SealMD5sum in SSE-S3 we need to protect the md5sum with compression - if we need to be fully S3 compatible. 
- If not test again by adding `-1` prefix at the end. 
- Server-side copy object is a legitimate problem that needs to be fixed, looks like that leads to a corrupted object.

## Steps to Reproduce (for bugs)
Just run mint tests against a setup like this
```yaml

<!--- Provi## Run minio dist erasure StatefulSet.
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: minio-compress-encrypt-dist-erasure-11445-f14a808
spec:
  serviceName: "minio-compress-encrypt-dist-erasure-headless-11445-f14a808"
  replicas: 4
  selector:
    matchLabels:
      app: minio-compress-encrypt-dist-erasure-app-11445-f14a808
  template:
    metadata:
      labels:
        app: minio-compress-encrypt-dist-erasure-app-11445-f14a808
    spec:
      containers:
      - name: minio-compress-encrypt-dist-erasure-11445-f14a808
        env:
        - name: MINIO_ACCESS_KEY
          value: "minio"
        - name: MINIO_SECRET_KEY
          value: "minio123"
        - name: MINIO_COMPRESS
          value: "true"
        - name: MINIO_COMPRESS_MIMETYPES
          value: "*"
        - name: MINIO_KMS_MASTER_KEY
          value: "my-minio-key:6368616e676520746869732070617373776f726420746f206120736563726574"
        image: minio/minio:11445-f14a808
        imagePullPolicy: Always
        args:
        - server
        - http://minio-compress-encrypt-dist-erasure-11445-f14a808-{0...3}.minio-compress-encrypt-dist-erasure-headless-11445-f14a808.default.svc.cluster.local/data
        ports:
        - containerPort: 9000
        resources:
          requests:
            cpu: 200m
            memory: 256Mi
            ephemeral-storage: 1024M
```

## Context
Compression and encryption should work seamlessly

## Regression
No

## Your Environment
All environments, latest release and master branch all have these bugs.

**Comment by harshavardhana** [comment]
This is still being worked on

**Comment by harshavardhana** [comment]
@aead @klauspost are you still planning to send PRs for the remaining failures in mint ?

**Comment by harshavardhana** [comment]
This is fully fixed after #12466 - closing no further action needed.
